### DISCUSSION TITLE
Design Patterns / App Architecture

### DISCUSSION THREAD

**Discussion by siddhesh12** [MobileNativeFoundation/discussions#42] [General]
Hey all, 
I'm curious to know what design patterns are used for developing large scale apps. MVC / MVVM / VIPER / FACTORY / CLEAN etc ? General strategies when it comes to choosing between design patterns?

**Comment by PeqNP** [comment]
One of the best (but most ambiguous) pieces of advice I have gotten is that your app needs to be "fit." Where it can be shaped and molded over time to adapt to the needs of the customers quickly. The question is, how can you make your app fit (robust, readable, changeable, etc.)?

Before I answer that question, in my experience 90% of the code can follow one pattern while other pages may require special attention. It doesn't matter if you use MVC / MVVM / VIPER / etc. for 90% of these pages. All of these should be perfectly suitable for most problems. The main challenge is discovering when to use a pattern that fits a more complex scenario.

For example, at Nordstrom most of our VC contain very little logic and send a "view event" signal to a model (`didLoadView`, `didTapCreateButton`, etc.). The model contains the business logic and prepares view states for the VC to consume. It sends a "command" signal back to the VC to update views, route to different parts of the app, etc. (`loadViewState`, `routeToCategoryResults`, etc.) VC -> M (& view state factories) -> VC. It's a fairly simple pattern that fit's more than 90% of our app's use cases. The model may contain a view state factory for each view in the VC and then composed within a single view state.

In the contexts where this pattern falls apart (many view event and command signals AND/OR where a large amount of state must be kept inside the model) we opt for a different pattern. In these cases we may use Rx for the signaling and create stateless models that can transform the state based on actions.

Some VCs we forego all patterns entirely and have the business logic in the VC, as the complexity of the logic doesn't warrant a model.

To answer this question directly:
> General strategies when it comes to choosing between design patterns?

If it's not obvious what pattern you should use, I would recommend liberally applying the principle of dependency inversion (protocol-oriented programming) and single responsibility. If you can understand and apply these two principles well, it is easy to identify emerging patterns and adapting your signaling pattern to best fit the context.

Does it go without saying that, regardless, you should be learning as many of these patterns as possible and practicing with them :)?

**Comment by sberrevoets** [comment]
The closest thing to a design pattern we have/had at Lyft was MVC. We used two strategies to avoid massive view controllers: protocol-oriented programming by putting a bunch of business logic in protocol extensions as "composition", and liberal use of child view controllers. The biggest problems with those approaches were:

- Lack of testability, which in the hyper growth phase wasn't all that useful since stuff would change too much to benefit from big test suites, but once things stabilized became a bigger problem we didn't address soon enough.
- The way we intended to use protocols was good, but with a rapidly growing team it was hard to stay consistent and keep everyone informed of best practices. This led to code becoming too intertwined while it looked like it was isolated and reusable. Nowadays we mostly don't use this pattern anymore.

In 2018 we defined what the backbone of our apps should look like—our "flow" architecture composes different user flows (like onboarding, payment, ride requesting, etc.) together to form an app and was inspired by [Square Workflow](https://square.github.io/workflow/) and redux. These flows are reusable across apps, have a unidirectional data flow, and are highly testable. Within those flows, developers had the freedom to build their features however they wanted.

We really liked the unidirectional patterns and what effect they had on the codebase so we're now also working to standardize their use in smaller features. Combined with our declarative UI layer on top of UIKit—giving us a SwiftUI-like syntax without needing to increase our deployment target—feature development is set to see some huge improvements this year and next.
